server/live_query/redis_live_query.go

```diff
@@ -167,63 +187,84 @@ func migrateBatchKeys(pool fleet.RedisPool, keys []string) error {
 	if len(hostIDs) == 0 {
 		return errors.New("no hosts targeted")
 	}
 
-	conn := r.pool.Get()
-	defer conn.Close()
-
-	// Map the targeted host IDs to a bitfield. Store targets in one key and SQL
-	// in another.
-	targetKey, sqlKey := generateKeys(name)
-	targets := mapBitfield(hostIDs)
-
-	// Ensure to set SQL first or else we can end up in a weird state in which a
-	// client reads that the query exists but cannot look up the SQL.
-	err := conn.Send("SET", sqlKey, sql, "EX", queryExpiration.Seconds())
-	if err != nil {
-		return fmt.Errorf("set sql: %w", err)
+	// store the sql and targeted hosts information
+	if err := r.storeQueryInfo(name, sql, hostIDs); err != nil {
+		return fmt.Errorf("store query info: %w", err)
 	}
-	_, err = conn.Do("SET", targetKey, targets, "EX", queryExpiration.Seconds())
-	if err != nil {
-		return fmt.Errorf("set targets: %w", err)
+
+	// store name (campaign id) into the active live queries set
+	if err := r.storeQueryNames(name); err != nil {
+		return fmt.Errorf("store query name: %w", err)
 	}
 
 	return nil
 }
 
 func (r *redisLiveQuery) StopQuery(name string) error {
-	conn := redis.ConfigureDoer(r.pool, r.pool.Get())
-	defer conn.Close()
+	// remove the sql and targeted hosts keys
+	if err := r.removeQueryInfo(name); err != nil {
+		return fmt.Errorf("remove query info: %w", err)
+	}
 
-	targetKey, sqlKey := generateKeys(name)
-	if _, err := conn.Do("DEL", targetKey, sqlKey); err != nil {
-		return fmt.Errorf("del query keys: %w", err)
+	// remove name (campaign id) from the livequery set
+	if err := r.removeQueryNames(name); err != nil {
+		return fmt.Errorf("remove query name: %w", err)
 	}
 
 	return nil
 }
 
+// this is a variable so it can be changed in tests
+var cleanupExpiredQueriesModulo int64 = 10
+
 func (r *redisLiveQuery) QueriesForHost(hostID uint) (map[string]string, error) {
 	// Get keys for active queries
-	queryKeys, err := redis.ScanKeys(r.pool, queryKeyPrefix+"*", 100)
+	names, err := r.loadActiveQueryNames()
 	if err != nil {
-		return nil, fmt.Errorf("scan active queries: %w", err)
+		return nil, fmt.Errorf("load active queries: %w", err)
 	}
 
-	keysBySlot := redis.SplitKeysBySlot(r.pool, queryKeys...)
+	// convert the query name (campaign id) to the key name
+	for i, name := range names {
+		tkey, _ := generateKeys(name)
+		names[i] = tkey
+	}
+
+	keysBySlot := redis.SplitKeysBySlot(r.pool, names...)
 	queries := make(map[string]string)
+	expired := make(map[string]struct{})
 	for _, qkeys := range keysBySlot {
-		if err := r.collectBatchQueriesForHost(hostID, qkeys, queries); err != nil {
+		if err := r.collectBatchQueriesForHost(hostID, qkeys, queries, expired); err != nil {
 			return nil, err
 		}
 	}
+
+	if len(expired) > 0 {
+		// a certain percentage of the time so that we don't overwhelm redis with a
+		// bunch of similar deletion commands at the same time, clean up the
+		// expired queries.
+		if time.Now().UnixNano()%cleanupExpiredQueriesModulo == 0 {
```

Comment:
I went with a fixed modulo value here, I'm not sure it would make sense to expose as a configuration option as it would be quite tricky to explain and for the user to identify a proper value.